tikv-client/src/main/java/com/pingcap/tikv/predicates/TiKVScanAnalyzer.java

```diff
@@ -305,62 +317,176 @@ TiKVScanPlan buildIndexScan(
       checkArgument(
           !ir.hasRange(), "Table scan must have one and only one access condition / point");
 
       Key key = ir.getAccessKey();
       checkArgument(key instanceof TypedKey, "Table scan key range must be typed key");
       TypedKey typedKey = (TypedKey) key;
-      startKey = RowKey.toRowKey(id, typedKey);
-      endKey = startKey.next();
+      signedStartKey = RowKey.toRowKey(id, typedKey);
+      signedEndKey = signedStartKey.next();
     } else if (ir.hasRange()) {
       checkArgument(
           !ir.hasAccessKey(), "Table scan must have one and only one access condition / point");
       Range<TypedKey> r = ir.getRange();
 
       if (!r.hasLowerBound()) {
         // -INF
-        startKey = RowKey.createMin(id);
+        signedStartKey = RowKey.createMin(id);
       } else {
         // Comparison with null should be filtered since it yields unknown always
-        startKey = RowKey.toRowKey(id, r.lowerEndpoint());
+        signedStartKey = RowKey.toRowKey(id, r.lowerEndpoint());
         if (r.lowerBoundType().equals(BoundType.OPEN)) {
-          startKey = startKey.next();
+          signedStartKey = signedStartKey.next();
         }
       }
 
       if (!r.hasUpperBound()) {
         // INF
-        endKey = RowKey.createBeyondMax(id);
+        signedEndKey = RowKey.createBeyondMax(id);
       } else {
-        endKey = RowKey.toRowKey(id, r.upperEndpoint());
+        signedEndKey = RowKey.toRowKey(id, r.upperEndpoint());
         if (r.upperBoundType().equals(BoundType.CLOSED)) {
-          endKey = endKey.next();
+          signedEndKey = signedEndKey.next();
         }
       }
     } else {
       throw new TiClientInternalException("Empty access conditions");
     }
-    return new Pair<>(startKey, endKey);
+    return new ScanRange(signedStartKey, signedEndKey, null, null);
+  }
+
+  private ScanRange buildTableScanUnsignedKeyRangePerId(long id, IndexRange ir) {
+    Key signedStartClosedKey;
+    Key signedEndOpenKey;
+    Key unsignedStartClosedKey;
+    Key unsignedEndOpenKey;
+
+    if (ir.hasAccessKey()) {
+      checkArgument(
+          !ir.hasRange(), "Table scan must have one and only one access condition / point");
+      Key key = ir.getAccessKey();
+      checkArgument(key instanceof TypedKey, "Table scan key range must be typed key");
+      BigDecimal value = (BigDecimal) ((TypedKey) key).getValue();
+      if (value.compareTo(MAX_SIGNED_LONG) > 0) {
+        unsignedStartClosedKey = RowKey.toRowKey(id, new IntHandle(value.longValue()));
+        unsignedEndOpenKey = unsignedStartClosedKey.nextPrefix();
+        return new ScanRange(null, null, unsignedStartClosedKey, unsignedEndOpenKey);
+      } else {
+        signedStartClosedKey = RowKey.toRowKey(id, new IntHandle(value.longValue()));
+        signedEndOpenKey = signedStartClosedKey.nextPrefix();
+        return new ScanRange(signedStartClosedKey, signedEndOpenKey, null, null);
+      }
+    } else if (ir.hasRange()) {
```

Comment:
we can just use `if` rather than else if because we have returned

Reply:
changed